src/libraries/System.Security.Cryptography.Cose/src/System/Security/Cryptography/Cose/CoseMessage.cs

```diff
@@ -0,0 +1,184 @@
+                throw new CryptographicException(SR.Format(SR.DecodeSign1ErrorWhileDecoding, SR.DecodeSign1EncodedProtectedMapIncorrect));
+            }
+            else if (protectedHeaderAsBstr.Length == 1 && protectedHeaderAsBstr[0] == EmptyStringByte)
+            {
+                return;
+            }
+
+            var protectedHeaderReader = new CborReader(protectedHeaderAsBstr);
+            DecodeBucket(protectedHeaderReader, headerParameters);
+
+            if (protectedHeaderReader.BytesRemaining != 0)
+            {
+                throw new CryptographicException(SR.Format(SR.DecodeSign1ErrorWhileDecoding, SR.DecodeSign1EncodedProtectedMapIncorrect));
+            }
+        }
+
+        private static void DecodeUnprotectedBucket(CborReader reader, CoseHeaderMap headerParameters)
+        {
+            DecodeBucket(reader, headerParameters);
+        }
+
+        private static void DecodeBucket(CborReader reader, CoseHeaderMap headerParameters)
+        {
+            int? length = reader.ReadStartMap();
+            for (int i = 0; i < length; i++)
+            {
+                CoseHeaderLabel label = reader.PeekState() switch
+                {
+                    CborReaderState.UnsignedInteger or CborReaderState.NegativeInteger => new CoseHeaderLabel(reader.ReadInt32()),
+                    CborReaderState.TextString => new CoseHeaderLabel(reader.ReadTextString()),
+                    _ => throw new CryptographicException(SR.Format(SR.DecodeSign1ErrorWhileDecoding, SR.DecodeSign1MapLabelWasIncorrect))
+                };
+                headerParameters.SetEncodedValue(label, reader.ReadEncodedValue());
+            }
+            reader.ReadEndMap();
+        }
+
+        private static byte[]? DecodePayload(CborReader reader)
+        {
+            CborReaderState state = reader.PeekState();
+            if (state == CborReaderState.Null)
+            {
+                reader.ReadNull();
+                return null;
+            }
+            if (state == CborReaderState.ByteString)
+            {
+                return reader.ReadByteString();
+            }
+
+            throw new CryptographicException(SR.Format(SR.DecodeSign1ErrorWhileDecoding, SR.DecodeSign1PayloadWasIncorrect));
+        }
+
+        private static byte[] DecodeSignature(CborReader reader)
+        {
+            return reader.ReadByteString();
+        }
+
+        internal static byte[] CreateToBeSigned(string context, ReadOnlySpan<byte> encodedProtectedHeader, ReadOnlySpan<byte> content)
+        {
+            var writer = new CborWriter();
+            writer.WriteStartArray(4);
+            writer.WriteTextString(context); // context
+            writer.WriteByteString(encodedProtectedHeader); // body_protected
+            writer.WriteByteString(Span<byte>.Empty); // external_aad
+            writer.WriteByteString(content); //payload or content
+            writer.WriteEndArray();
+            return writer.Encode();
```

Comment:
This is fine for now, but if we decide to take the partner feedback of allowing detached content to be presented as a stream (to support large detached content) we'll want to replace this with just a virtual data structure that we pipe through IncrementalHash.